_posts/2018-12-10-blog-lets-create some-glip-bots-with-python.markdown

```diff
@@ -0,0 +1,321 @@
+RINGCENTRAL_BOT_SERVER=https://xxxxx.ngrok.io
+
+## for bots auth required, get them from your ringcentral app page
+RINGCENTRAL_BOT_CLIENT_ID=
+RINGCENTRAL_BOT_CLIENT_SECRET=
+
+# create custom bot config file
+# since all functions are optional, you could keep those you edited, delete others in config.py
+cp config.sample.py config.py
+
+# run local dev server
+./bin/start
+```
+
+## Test bot
+
+- Goto your ringcentral app's bot section, click 'Add to glip'
+- Login to [https://glip-app.devtest.ringcentral.com](https://glip-app.devtest.ringcentral.com), find the bot by searching its name. Talk to the bot.
+- Edit config.py to change bot bahavior and test in [https://glip-app.devtest.ringcentral.com](https://glip-app.devtest.ringcentral.com)
+
+## Let's write some bot logic
+
+let's write a bot that can tell server time/date as a hello world bot
+
+Edit `config.py`, like this:
+
+```py
+"""
+sample config module
+run "cp config.sample.py config.py" to create local config
+edit config.py functions to override default bot behavior
+"""
+
+from datetime import datetime
+
+__name__ = 'localConfig'
+__package__ = 'ringcentral_bot_framework'
+
+def helpMsg(botId):
+  return f'''Hello, I am a date/time chatbot. Please reply "@![:Person]({botId}) **cmd**" if you want to talk to me.
+
+**cmd** list
+
+**date** -- show current date
+**time** -- show current time
+  '''
+
+def botJoinPrivateChatAction(bot, groupId, user, dbAction):
+  """
+  bot join private chat event handler
+  bot could send some welcome message or help, or something else
+  """
+  text = helpMsg(bot.id)
+  bot.sendMessage(
+    groupId,
+    {
+      'text': text
+    }
+  )
+
+def botGotPostAddAction(
+  bot,
+  groupId,
+  creatorId,
+  user,
+  text,
+  dbAction,
+  handledByExtension
+):
+  """
+  bot got group chat message: text
+  bot could send some response
+  """
+  if handledByExtension:
+    return
+
+  if text == f'![:Person]({bot.id}) date':
+    date = str(datetime.now().strftime('%Y-%m-%d'))
+    bot.sendMessage(
+      groupId,
+      {
+        'text': f'![:Person]({creatorId}), current date is {date}'
+      }
+    )
+  elif text == f'![:Person]({bot.id}) time':
+    time = str(datetime.now().strftime('%Y-%m-%d %H:%M:%S'))
+    bot.sendMessage(
+      groupId,
+      {
+        'text': f'![:Person]({creatorId}), current time is {time}'
+      }
+    )
+  elif f'![:Person]({bot.id})' in text:
+    bot.sendMessage(
+      groupId,
+      {
+        'text': helpMsg(bot.id)
+      }
+    )
+
+
+```
+
+Then try talk to the bot in [https://glip-app.devtest.ringcentral.com](https://glip-app.devtest.ringcentral.com), see how it work.
+
+That is it. You can just focus on the bot logic.
+
+If you want to access user data, subscribe to user event, or use different database, just check these examples.
+
+## Example bots
+
+- [date-time-chatbot](https://github.com/zxdong262/ringcentral-date-time-chatbot) : Simple ringcentral chatbot which can tell time/date.
```

Comment:
Simple RingCentral chatbot 'that' can tell time/date.
Use proper case for company names, especially our own company name!
Usually if it's in one phrase, use 'that', only use 'which' if you are breaking down the sentence with comma(',').
For example: I made a bot "that" can check my schedule. It is a bot made in python, which is very easy to setup.